ssf/samples_test.go

```diff
@@ -123,3 +123,18 @@ func BenchmarkRandomlySample(b *testing.B) {
 	b.ResetTimer()
 	samples = RandomlySample(0.2, samples...)
 }
+
+func BenchmarkRandomlySampleParallel(b *testing.B) {
+
+	samples := make([]*SSFSample, 1000000)
+	for i := range samples {
+		samples[i] = Count("testing.counter", float32(i), nil)
+	}
```

Comment:
One thing I overlooked, this benchmark just `panic`ed when I ran it on my machine: `RandomlySample`ing this data set isn't parallelism-safe since `RandomlySample` changes the SampleRate on each sample.

Reply:
Here's the panic trace:

```
$ go test -run None -bench Bench ./ssf
goos: darwin
goarch: amd64
pkg: github.com/stripe/veneur/ssf
BenchmarkRandomlySample-4               2000000000               0.02 ns/op
BenchmarkRandomlySampleParallel-4       panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x3c pc=0x126ba1f]

goroutine 37 [running]:
github.com/stripe/veneur/ssf.RandomlySample(0x3e4ccccd, 0x1525528, 0x9b8c, 0x30c02, 0x0, 0x0, 0x0)
        /Users/asf/go/src/github.com/stripe/veneur/ssf/samples.go:132 +0x16f
github.com/stripe/veneur/ssf.BenchmarkRandomlySampleParallel.func1(0xc4200a60c0)
        /Users/asf/go/src/github.com/stripe/veneur/ssf/samples_test.go:137 +0x9f
testing.(*B).RunParallel.func1(0xc420166020, 0xc420166018, 0xc420166010, 0xc420162c00, 0xc42008a1c0)
        /usr/local/Cellar/go/1.10/libexec/src/testing/benchmark.go:623 +0x98
created by testing.(*B).RunParallel
        /usr/local/Cellar/go/1.10/libexec/src/testing/benchmark.go:616 +0x199
exit status 2
FAIL    github.com/stripe/veneur/ssf    3.703s
```

Reply:
I think this should fix the problem, at the cost of reaching deep into the guts of the runtime to do so /:

``` go
func BenchmarkRandomlySampleParallel(b *testing.B) {
	var proc uint64
	nSamples := 1000000
	allSamples := make([]*SSFSample, nSamples*runtime.GOMAXPROCS(-1)) // "The number of goroutines [from b.RunParallel] defaults to GOMAXPROCS"
	for i := range allSamples {
		allSamples[i] = Count("testing.counter", float32(i), nil)
	}
	b.ResetTimer()

	b.RunParallel(func(pb *testing.PB) {
		samples := allSamples[int(proc)*nSamples : (int(proc)+1)*nSamples-1]
		atomic.AddUint64(&proc, 1)
		for pb.Next() {
			samples = RandomlySample(0.2, samples...)
		}
	})
}
```

Reply:
I don't *think* we want to use an atomic here, because that's reintroducing a serialization point in the code (and we're explicitly trying to test the behavior in parallel). However, we can avoid this race condition (see below)